src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,325 @@
+
+    # prepare for basis calculation
+    ch = xp.exp(-tl[:, None, ...] @ zk[None, ...])
+    w = xp.diag(hk**0.5)
+    p = xp.linalg.pinv(w @ _transpose(ch)) @ w
+
+    # actual temporal basis calculation
+    B = p @ xp.exp(-zk[:, None, ...] * t[None, ...])
+    B = B.astype(xp.complex64)
+
+    # get spatial coeffs
+    C = xp.exp(-tl * fieldmap[..., None])
+    C = C[None, ...].swapaxes(0, -1)[
+        ..., 0
+    ]  # (..., n_time_segments) -> (n_time_segments, ...)
+
+    # clean-up of spatial coeffs
+    C = xp.nan_to_num(C, nan=0.0, posinf=0.0, neginf=0.0)
+
+    # back to torch if required
+    if is_torch:
+        if xp.__name__ == "cupy":
+            B = torch.from_dlpack(B)
+            C = torch.from_dlpack(C)
+        else:
+            B = torch.from_numpy(B)
+            C = torch.from_numpy(C)
+
+    return B, C
+
+
+def _outer_sum(xx, yy):
+    xx = xx[:, None, ...]  # add a singleton dimension at axis 1
+    yy = yy[None, ...]  # add a singleton dimension at axis 0
+    ss = xx + yy  # compute the outer sum
+    return ss
+
+
+def _transpose(input):
+    xp = get_array_module(input)
+    if xp.__name__ == "torch":
+        return input.mT
+    else:
+        return input.T
+
+
+class MRIFourierCorrected(FourierOperatorBase):
+    """Fourier Operator with B0 Inhomogeneities compensation.
+
+    This is a wrapper around the Fourier Operator to compensate for the
+    B0 inhomogeneities  in the  k-space.
+
+    Parameters
+    ----------
+    fourier_op: object of class FourierBase
+        the fourier operator to wrap
+    fieldmap : np.ndarray or GPUarray, optional
+        Rate map defined as ``fieldmap = R2*_map + 1j * B0_map``.
+        ``*_map`` and ``t`` should have reciprocal units.
+        If ``zmap`` is real, assume ``zmap = B0_map``.
+        Expected shape is ``(nz, ny, nx)``.
+    t : np.ndarray or GPUarray, optional
+        Readout time in ``[s]`` of shape ``(nshots, npts)`` or ``(nshots * npts,)``.
+    n_time_segments : int, optional
+        Number of time segments. The default is ``6``.
+    n_bins : int | Sequence[int] optional
+        Number of histogram bins to use for ``(B0, T2*)``. The default is ``(40, 10)``
+        If it is a scalar, assume ``n_bins = (n_bins, 10)``.
+        For real fieldmap (B0 only), ``n_bins[1]`` is ignored.
+    mask : np.ndarray or GPUarray, optional
+        Boolean mask to avoid histogram of background values.
+        The default is ``None`` (use the whole map).
+    B : np.ndarray or GPUarray, optional
+        Temporal interpolator of shape ``(n_time_segments, len(t))``.
+    C : np.ndarray or GPUarray, optional
+        Off-resonance phase map at each time segment center of shape
+        ``(n_time_segments, *fieldmap.shape)``.
+    backend: str, optional
+        The backend to use for computations. Either 'cpu', 'gpu' or 'torch'.
+        The default is `cpu`.
+    """
+
+    def __init__(
+        self,
+        fourier_op,
+        fieldmap=None,
+        t=None,
```

Comment:
Consider calling it `time` for better understanding.

Reply:
Do you think `readout_time` would be too verbose?

Reply:
No I think for parameters it's ok to be verbose